network/ssh.go

```diff
@@ -0,0 +1,137 @@
+/*
+   Copyright 2015 CoreOS, Inc.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package network
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/coreos/coreos-cloudinit/config"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/agent"
+)
+
+const (
+	defaultUser = "core"
+	rsaKeySize  = 2048
+)
+
+// An interface for anything compatible with net.Dialer
+type Dialer interface {
+	Dial(network, address string) (net.Conn, error)
+}
+
+// SSHAgent can manage keys, updates cloud config, and loves ponies.
+// The embedded dialer is used for establishing new SSH connections.
+type SSHAgent struct {
+	agent.Agent
+	Dialer
```

Comment:
network.Dialer adds support for retrying on failure. A helpful property when waiting for a machine to boot.